vowpalwabbit/learner.h

```diff
@@ -128,6 +134,173 @@ inline void decrement_offset(multi_ex& ec_seq, const size_t increment, const siz
   }
 }
 
+#define TYPED_EXAMPLE_LEARN_FN(TYPE, FIELD_NAME)                                                                       \
+  inline void learn_with_label(example& ec, TYPE& label, size_t i = 0)                                                 \
+  {                                                                                                                    \
+    assert(!is_multiline);                                                                                             \
+    increment_offset(ec, increment, i);                                                                                \
+    if (learn_fd.learn_with_label_f != nullptr)                                                                        \
+    {                                                                                                                  \
+      learn_fd.learn_with_label_f(learn_fd.data, *learn_fd.base, static_cast<void*>(&ec), static_cast<void*>(&label)); \
+    }                                                                                                                  \
+    else                                                                                                               \
+    {                                                                                                                  \
+      polylabel saved = std::move(ec.l);                                                                               \
+      ec.l.FIELD_NAME = std::move(label);                                                                              \
```

Comment:
>FIELD_NAME [](start = 11, length = 10)

These macros mostly seem to have to do with the fact that the name of the variable in label is tied to the label type.  Is that right?
If that is the case, I am wondering if this can be accomplished through some form of template specialization around just this bit of code

Reply:
The union must be accessed using the named field, as reinterpret casting is breaks if it becomes a struct. I'd like to avoid that debt as I'm all too familiar with it. (Even though this code will be removed by then). Given that you'd need to overload on every type and at that point your back to the same amount of duplication here